### PR TITLE
Implement strtok

### PR DESCRIPTION
**Pull request by tdbgamer**
I implemented strtok, but it uses some of the functions whose signatures will be changed for issue #89. Once that one merges, this one should be good to go unless you see and mistakes.

### CODE REVIEW

File: src/string/src/lib.rs [hunk 1]
```diff
@@ -323,21 +323,79 @@ pub unsafe extern "C" fn strstr(s1: *const c_char, s2: *const c_char) -> *mut c_
...
+            if (*lasts).is_null() {
+                return ptr::null_mut();
+            }
+            haystack = *lasts;
+        }
+
+        // Skip past any extra delimiter left over from previous call
+        haystack = haystack.add(strspn(haystack, delimiter));
+        if *haystack == 0 {
+            haystack = ptr::null_mut();
```
Comment (dlrobertson) [inline src/string/src/lib.rs]: `haystack` isn't a static, so you don't need to assign `NULL` to it right?

File: src/string/src/lib.rs [hunk 2]
```diff
@@ -323,21 +323,79 @@ pub unsafe extern "C" fn strstr(s1: *const c_char, s2: *const c_char) -> *mut c_
...
+
+        // Build token by injecting null byte into delimiter
+        let token = haystack;
+        haystack = strpbrk(token, delimiter);
+        if !haystack.is_null() {
+            haystack.write(0);
+            haystack = haystack.add(1);
+            *lasts = haystack;
+        } else {
+            haystack = ptr::null_mut();
```
Comment (dlrobertson) [inline src/string/src/lib.rs]: Same

**Comment by tdbgamer** [comment]
Ah right, good catch, I'll fix it real quick

File: src/string/src/lib.rs [hunk 3]
```diff
@@ -323,21 +323,77 @@ pub unsafe extern "C" fn strstr(s1: *const c_char, s2: *const c_char) -> *mut c_
 }
 
 #[no_mangle]
-pub extern "C" fn strtok(s1: *mut c_char, s2: *const c_char) -> *mut c_char {
-    unimplemented!();
+pub extern "C" fn strtok(s1: *mut c_char, delimiter: *const c_char) -> *mut c_char {
```
Comment (ids1024) [inline src/string/src/lib.rs]: Would it be possible to replace this with a short function that just calls `strtok_r`, passing it a static?

Comment (tdbgamer) [reply src/string/src/lib.rs]: Sure that's probably fine. I'll test it out later.

**Comment by tdbgamer** [comment]
all good now?

**Comment by jackpot51** [comment]
Thanks!

**Pull request merged**